File: neetcode/array-hashing/duplicate_integer.py
# from sortedcontainers import SortedList
from typing import List


class Solution:
    """
        https://neetcode.io/problems/duplicate-integer
    """
    def hasDuplicate(self, nums: List[int]) -> bool:
        hashes = {}
        for n in nums:
            hashes.update({n: hashes.get(n, 0) + 1})
            if hashes[n] > 1:
                return True

        return False


nums = [1, 2, 3, 3]

solution = Solution()
print(solution.hasDuplicate([1, 2, 3, 3]))
print(solution.hasDuplicate([1, 2, 3]))
print(solution.hasDuplicate([3, 3, 1, 2]))

File: neetcode/array-hashing/duplicate-integer.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
        unordered_map<int, int> nums_map;
        for (int i = 0; i < nums.size(); i++)
        {
            nums_map[nums[i]]++;
            if (nums_map[nums[i]] > 1)
            {
                return true;
            }
        }
        return false;
    }
};

int main() {
    Solution s = Solution();

    vector<int> nums = {1, 2, 3, 3};
    cout << s.hasDuplicate(nums) << endl;

    nums = {1, 3, 2, 3};
    cout << s.hasDuplicate(nums) << endl;

    nums = {1, 2, 3};
    cout << s.hasDuplicate(nums) << endl;

    nums = {1, 2, 3, 4};
    cout << s.hasDuplicate(nums) << endl;

    return 0;
}
